### PR TITLE
User Can Login

### PR DESCRIPTION
**Pull request by pedrocbras**
Testing for Log in feature with fake user
Testing for Success message result
PT Log in: https://www.pivotaltracker.com/story/show/167062799
PT Success Message: https://www.pivotaltracker.com/story/show/167062881

### CODE REVIEW

File: features/log_in.feature [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    Given the following user exists
+      | email        | password   | name |
+      | john@doe.com | password1  | John |
+    Given I visit the site
+    Then I should see "Login"
+    When I click "Login"
+    Then I fill in "Email" with "john@doe.com"
+    And I fill in "Password" with "password1"
+    And I click "Log in"
+    Then i should see "Signed in successfully."
```
Comment (faraznaeem) [inline features/log_in.feature]: Capitalize `I`

File: features/step_definitions/log_in_steps.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
+Given("the following user exists") do |table|
+    table.hashes.each do |table|
+      User.create(table)
```
Comment (faraznaeem) [inline features/step_definitions/log_in_steps.rb]: Please use a factory here

File: features/step_definitions/log_in_steps.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
+Given("the following user exists") do |table|
+    table.hashes.each do |table|
```
Comment (faraznaeem) [inline features/step_definitions/log_in_steps.rb]: please use another iteration variable like `user` to properly reflect the iteration

File: features/step_definitions/log_in_steps.rb [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+      User.create(table)
+    end
+  end 
+
+  Given("I visit the site") do
+    visit root_path
```
Comment (faraznaeem) [inline features/step_definitions/log_in_steps.rb]: Please either use brackets or not, just stay consistent

File: features/step_definitions/log_in_steps.rb [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  Then("I should see {string}") do |button|
+    expect(page).to have_content(button)
+  end
+  
+  When("I click {string}") do |login|
+    click_on(login)
+  end
+  
+  Then("I fill in {string} with {string}") do |field, content|
+    fill_in(field, :with => content)
```
Comment (faraznaeem) [inline features/step_definitions/log_in_steps.rb]: Please use new ruby syntax and not hash rockets `=>` as mentioned in yesterdays demo

File: features/step_definitions/log_in_steps.rb [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+  end
+  
+  Then("I click on {string}") do |log_in|
+    click_on(log_in)
+  end
+  
+  Then("i should see {string}") do |success|
+    expect(page).to have_content(success)
```
Comment (faraznaeem) [inline features/step_definitions/log_in_steps.rb]: Will it always be `success`? The better practice is to write the iteration variable as something that can be reused and reflect more generic cases.

File: features/step_definitions/log_in_steps.rb [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+  When("I click {string}") do |login|
+    click_on(login)
+  end
+  
+  Then("I fill in {string} with {string}") do |field, content|
+    fill_in(field, :with => content)
+  end
+  
+  Then("I click on {string}") do |log_in|
+    click_on(log_in)
```
Comment (faraznaeem) [inline features/step_definitions/log_in_steps.rb]: Will it always be `log_in`? The better practice is to write the iteration variable as something that can be reused and reflect more generic cases.

**Comment by erikwesterberg** [comment]
Added sad path to the feature test!

File: features/log_in.feature (hunk 1 above)
Comment (erikwesterberg) [reply features/log_in.feature]: fixed

File: features/step_definitions/log_in_steps.rb (hunk 2 above)
Comment (erikwesterberg) [reply features/step_definitions/log_in_steps.rb]: fixed

File: features/step_definitions/log_in_steps.rb (hunk 5 above)
Comment (erikwesterberg) [reply features/step_definitions/log_in_steps.rb]: fixed

File: features/log_in.feature [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+Scenario: If user wants to log in
+    Then I should see "Login"
+    When I click "Login"
+    Then I fill in "Email" with "john@doe.com"
+    And I fill in "Password" with "password1"
+    And I click "Log in"
+    Then I should see "Signed in successfully."
+
+Scenario: If user fill in invalid answers
```
Comment (oliverochman) [inline features/log_in.feature]: Scenario is `User can login with invalid credentials [Sad Path]`

File: features/log_in.feature [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    And I click "Log in"
+    Then I should see "Invalid Email or password"
+    Then I fill in "Email" with ""
+    And I fill in "Password" with ""
+    And I click "Log in"
+    Then I should see "Invalid Email or password"
+
+
+Scenario: Not possible to log in if no info is given
+    When I click "Login"
```
Comment (oliverochman) [inline features/log_in.feature]: Refactor this like I asked you to do in the `Visitor can Sign up` PR

File: features/log_in.feature [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+    And I click "Log in"
+    Then I should see "Invalid Email or password."
+    When I fill in "Email" with ""
+    And I click "Log in"
+    Then I should see "Invalid Email or password."
+    When I fill in "Password" with ""
+    And I click "Log in"
+    Then I should see "Invalid Email or password."
+  
+Scenario: User fucking up with wrong password
```
Comment (oliverochman) [inline features/log_in.feature]: Keep it professional
Scenario is `User cant login with wrong password [Sad Path]`

File: features/log_in.feature [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+    Then I should see "Login"
+    When I click "Login"
+    And I fill in "Email" with "johndoe@gmail.com"
+    When I fill in "Password" with "234"
```
Comment (oliverochman) [inline features/log_in.feature]: change `"234"` to `"wrong_password"`

Comment (erikwesterberg) [reply features/log_in.feature]: Fixed

File: features/log_in.feature (hunk 10 above)
Comment (erikwesterberg) [reply features/log_in.feature]: fixed

File: features/log_in.feature (hunk 8 above)
Comment (erikwesterberg) [reply features/log_in.feature]: Fixed

File: features/log_in.feature [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+    When I enter my account
+    I need be able to log in
+
+Background:
+Given the following user exists
+      | email        | password   | name |
+      | john@doe.com | password1  | John |
+Given I visit the site
+
+Scenario: If user wants to log in
```
[the remark this answers is not in the document]
Comment (erikwesterberg) [reply features/log_in.feature]: fixed

File: features/log_in.feature [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+    Then I should see "Login"
+    When I click "Login"
+    Then I fill in "Email" with "john@doe.com"
```
[the remark this answers is not in the document]
Comment (erikwesterberg) [reply features/log_in.feature]: fixed

File: features/log_in.feature [hunk 14]
```diff
@@ -0,0 +1,52 @@
+Feature: Check for a valid email
+    As a user,
+    When I enter my account
+    I need be able to log in
+
+Background:
+Given the following user exists
+      | email        | password   | name |
+      | john@doe.com | password1  | John |
+Given I visit the site
```
[the remark this answers is not in the document]
Comment (erikwesterberg) [reply features/log_in.feature]: fixed

File: features/log_in.feature [hunk 15]
```diff
@@ -0,0 +1,52 @@
+Feature: Check for a valid email
+    As a user,
+    When I enter my account
+    I need be able to log in
+
+Background:
```
[the remark this answers is not in the document]
Comment (erikwesterberg) [reply features/log_in.feature]: Ill think u have to show me where

File: features/log_in.feature [hunk 16]
```diff
@@ -0,0 +1,52 @@
+Feature: Check for a valid email
```
[the remark this answers is not in the document]
Comment (erikwesterberg) [reply features/log_in.feature]: fixed in 1 minuto

File: features/log_in.feature (hunk 15 above)
[the remark this answers is not in the document]
Comment (oliverochman) [reply features/log_in.feature]: look at the indentation in the `visitor can sign up` feature file

File: features/log_in.feature [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+
+Background:
+Given the following user exists
+      | email        | password   | name |
+      | john@doe.com | password1  | John |
+Given I visit the site
+
+Scenario: If user wants to log in
+    Then I should see "Login"
+    When I click "Login"
```
[the remark this answers is not in the document]
Comment (erikwesterberg) [reply features/log_in.feature]: fixed

File: features/log_in.feature (hunk 15 above)
[the remark this answers is not in the document]
Comment (erikwesterberg) [reply features/log_in.feature]: fixed

File: features/log_in.feature [hunk 18]
```diff
@@ -0,0 +1,49 @@
+Feature: User can login
+    As a user,
+    When I enter my account
+    I need be able to log in
```
Comment (faraznaeem) [inline features/log_in.feature]:
```
As a user 
In order to access my account
I would like to be able to log in
```

File: features/step_definitions/log_in_steps.rb (hunk 3 above)
Comment (faraznaeem) [reply features/step_definitions/log_in_steps.rb]: please fix this

File: features/log_in.feature (hunk 18 above)
Comment (erikwesterberg) [reply features/log_in.feature]: Fixed

File: features/step_definitions/log_in_steps.rb (hunk 3 above)
Comment (erikwesterberg) [reply features/step_definitions/log_in_steps.rb]: Fixed, sorry for delay!

**Pull request merged**